src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Rename/FileMoveNotificationListener.cs

```diff
@@ -56,190 +58,141 @@ internal class FileMoveNotificationListener : IFileMoveNotificationListener
+                        IEnumerable<IFileMoveItem> children = copyPasteItem.Children.Select(c => c as IFileMoveItem).WhereNotNull();
+                        foreach (IFileMoveItem child in GetFilesToMoveRecursive(children))
+                        {
+                            yield return child;
+                        }
                     }
-
-                    actions.Add((documentAction, filename));
                 }
-
-                return actions;
             }
         }
 
         public async Task OnAfterFileMoveAsync()
         {
-            if (_actions is { Count: not 0 } && await CheckUserConfirmationAsync())
+            if (!_renameActionSets.Any() || !await IsEnabledOrConfirmedAsync())
             {
-                ApplyNamespaceUpdateActions();
+                return;
             }
 
-            return;
-
-            async Task<bool> CheckUserConfirmationAsync()
+            _ = _threadingService.JoinableTaskFactory.RunAsync(async () =>
             {
-                ISettingsManager settings = await _settingsManagerService.GetValueAsync();
+                // The wait service requires the main thread to run.
+                await _threadingService.SwitchToUIThread();
+                // Displays a dialog showing the progress of updating the namespaces in the files.
+                _waitService.Run(
+                    title: string.Empty,
+                    message: _renameMessage!,
+                    allowCancel: true,
+                    asyncMethod: ApplyRenamesAsync,
+                    totalSteps: _renameActionSets.Count);
+            });
 
-                bool promptNamespaceUpdate = settings.GetValueOrDefault(VsToolsOptions.OptionPromptNamespaceUpdate, true);
-                bool enabledNamespaceUpdate = settings.GetValueOrDefault(VsToolsOptions.OptionEnableNamespaceUpdate, true);
+            return;
 
-                if (!enabledNamespaceUpdate || !promptNamespaceUpdate)
+            async Task ApplyRenamesAsync(IWaitContext context)
+            {
+                CancellationToken token = context.CancellationToken;
+                await TaskScheduler.Default;
+                // WORKAROUND: We don't yet have a way to wait for the changes to propagate to Roslyn, tracked by https://github.com/dotnet/project-system/issues/3425
+                // Instead, we wait for the IntelliSense stage to finish for the entire solution.
+                IVsOperationProgressStatusService statusService = await _operationProgressService.GetValueAsync(token);
+                await statusService.GetStageStatus(CommonOperationProgressStageIds.Intellisense).WaitForCompletionAsync().WithCancellation(token);
+                // After waiting, a "new" published Solution is available.
+                Solution solution = _workspace.CurrentSolution;
+
+                int currentStep = 1;
+                foreach (KeyValuePair<string, RenameDocumentActionSet> renameActionSet in _renameActionSets)
```

Comment:
Consider deconstructing here:

```suggestion
                foreach ((string fileName, RenameDocumentActionSet renameActionSet) in _renameActionSets)
```

Reply:
I think I had tried to do this before, but it didn't work. I remember looking it up and saw that it was something specific to .NET Core, but you could get the package that included support for deconstructing Dictionary.
This was the response I read: https://stackoverflow.com/a/55392400/294804

However, looking at the actual marked answer: https://stackoverflow.com/a/42549535/294804

So, it seems to work because we have `System.ValueTuple`. The thing I'm confused by... I **know** VS showed an error before when I tried this because I wouldn't have searched for that StackOverflow post otherwise. Maybe the state of my file was in disarray so it couldn't resolve it properly... or it was a temporary bug we had in that particular VS build. Either way, this will be there now.

Reply:
We added an extension method to polyfill deconstruction for the type needed here (`KeyValuePair<,>`).

Reply:
Updated this to use deconstruction.